pylint/checkers/utils.py

```diff
@@ -1260,6 +1260,13 @@ def safe_infer(node: nodes.NodeNG, context=None) -> Optional[nodes.NodeNG]:
             inferred_type = _get_python_type_of_node(inferred)
             if inferred_type not in inferred_types:
                 return None  # If there is ambiguity on the inferred node.
+            if isinstance(inferred, nodes.FunctionDef):
+                if (
+                    inferred.args.args is not None
+                    and value.args.args is not None
+                    and len(inferred.args.args) != len(value.args.args)
+                ):
```

Comment:
```suggestion
                if len(inferred.args.args) != len(value.args.args):
```

Would this work? `nodes.Arguments.args` is typed as `typing.List[AssignName]` so it shouldn't ever be `None`?

Reply:
Maybe the typing in Astroid needs updating?

```
pylint/testutils/lint_module_test.py:181: AssertionError
---------------------------------- Captured stdout call ----------------------------------
Exception on node <Call l.10 at 0x7f7f171e7b20> in file '/Users/.../pylint/tests/functional/m/metaclass_attr_access.py'
---------------------------------- Captured stderr call ----------------------------------
Traceback (most recent call last):
  File "/Users/.../pylint/pylint/utils/ast_walker.py", line 72, in walk
    callback(astroid)
  File "/Users/.../pylint/pylint/checkers/refactoring/refactoring_checker.py", line 996, in visit_call
    self._check_consider_using_with(node)
  File "/Users/.../pylint/pylint/checkers/refactoring/refactoring_checker.py", line 1468, in _check_consider_using_with
    inferred = utils.safe_infer(node.func)
  File "/Users/.../pylint/pylint/checkers/utils.py", line 1264, in safe_infer
    if len(inferred.args.args) != len(value.args.args):
TypeError: object of type 'NoneType' has no len()
```

Reply:
This is actually a bug in `astroid`. See:
https://github.com/PyCQA/astroid/blob/6691a9e97373b28e15a76b747f78c041f2064df4/astroid/raw_building.py#L232
This should be:
`func.args.args = []`.

I tested and with this change the additional checks are unnecessary.
Let's block this until this is fixed in `astroid`. If you want you can make the PR there, I haven't come up with a test yet so you would need to do that yourself.

Diff:
```diff
diff --git a/astroid/raw_building.py b/astroid/raw_building.py
index 57e8d354f..e78dae7a2 100644
--- a/astroid/raw_building.py
+++ b/astroid/raw_building.py
@@ -229,7 +229,7 @@ def object_build_methoddescriptor(node, member, localname):
     )
     # set node's arguments to None to notice that we have no information, not
     # and empty argument list
-    func.args.args = None
+    func.args.args = []
     node.add_local_node(func, localname)
     _add_dunder_class(func, member)
```

Reply:
Wouldn't that introduce a behavior change here?

https://github.com/PyCQA/pylint/blob/3ed7a8e3933d14bc6d07afa8e6ba992b5a4b6926/pylint/checkers/classes.py#L1422

The code comments say it's trying to distinguish a truthful report of zero arguments from a total lack of information.

Reply:
Hmm, I have not looked at it well enough but this seems like something that should be fixed. It makes sense that `Arguments.args` is always initialised and allows removing checks such as the one in this PR. 
`astroid` probably shouldn't use that attribute itself to store information about the fact that it can't determine the attribute. If we had enabled `mypy` on `astroid` and had correct typing this wouldn't have been allowed anyway.

If I find the time I'll try and see if I can find a good solution for this.

Reply:
Pinging @cdce8p as you often have good ideas about how to represent things in `astroid`. Please disregard if you don't have the time to look into this!

The problem we're dealing with here is that for builtin functions or classes defined in C we can't get a signature and therefore can't get the arguments. In `astroid` we represent this by setting the `args` attribute of the `Arguments` class associated with that `FunctionDef` (set on `FunctionDef.args`) to `None`. See:
https://github.com/PyCQA/astroid/blob/6691a9e97373b28e15a76b747f78c041f2064df4/astroid/raw_building.py#L232

However, `Arguments.args` is typed as `typing.List[AssignName]`. Thus, the linked code is in direct violation of the current typing. Therefore, I don't think we should keep the status quo, but should try and find a solution.

Since there is no way to get the arguments we must indeed have a way to represent that the arguments simply can't be determined. This is also relevant information. For example, the fact that arguments can't be determined is used here:
https://github.com/PyCQA/pylint/blob/3ed7a8e3933d14bc6d07afa8e6ba992b5a4b6926/pylint/checkers/classes.py#L1422
I found about 3/4 other cases within `pylint` that also rely on this. I tried using a `node.parent.name == "builtins"` check, but that doesn't work for stuff like the `_io` module and other C-modules.

How to go forward? I think there are two options here:
1) Change the typing to become `typing.Optional[typing.List[AssignName]]` and lose the possibility of knowing you can iterate over `Arguments.args`.
2) Add a new attribute `missing_signature` (?) to `Arguments` and have the checks that rely on `args is not None` (or `is`) use that new attribute.

@jacobtylerwalls I'm also open to any of your suggestions for this problem. I propose to keep blocking this PR until this `astroid` issue has been resolved as it might impact the code of this PR.

Reply:
Tbh I wasn't aware of that behavior when I added the type annotation. Even if we wanted to untangle that into a new attribute, we couldn't without deprecating it first. As you mentioned, there is even code in pylint that depends on that behavior so it won't be unreasonable to assume that plugins might do too.

I would suggest we update the type annotation and docstring instead. It's the easiest thing to do here. Link me on the PR and I can take a look.

Reply:
See https://github.com/PyCQA/astroid/pull/1283